enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+to enable the user to pre-cache the additional images.
+
+The following changes are suggested:
+
+1. Add a new `AdditionalImages` field to the `PrecachingSpec` struct in
+`cluster-group-upgrades-operator/api/v1alpha1/clustergroupupgrade_types.go` as shown below.
+    ```golang
+    ...
+    // PrecachingSpec defines the pre-caching software spec derived from policies
+    type PrecachingSpec struct {
+        PlatformImage                string   `json:"platformImage,omitempty"`
+        OperatorsIndexes             []string `json:"operatorsIndexes,omitempty"`
+        OperatorsPackagesAndChannels []string `json:"operatorsPackagesAndChannels,omitempty"`
+        ExcludePrecachePatterns      []string `json:"excludePrecachePatterns,omitempty"`
+        AdditionalImages    []string `json:"additionalImages,omitempty"`
+    }
+   ...
+
+2. Apply the operator-sdk tool to update the TALM custom resource definitions, manifests and auto-generated
+golang source files.
+
+3. Extend the `controllers/managedClusterResources.go` and
+`cluster-group-upgrades-operator/controllers/precache.go` source files to handle reconciliation
+control logic pertaining to the `additionalImages` object.
+
+4. Extend the pre-cache template source file `controllers/templates/precache-templates.go` with the
+`additionalImages` field.
+
+5. Refactor the `cluster-group-upgrades-operator/pre-cache/precache.sh` and
+`cluster-group-upgrades-operator/pre-cache/pull` scripts to check for additional user images and pull
+them accordingly. The existing image pulling functionality should be refactored in order to promote
+code reusability and avoid code duplication.
+
+### Workflow Description
+
+The general workflow for pre-caching user specified images is as follows:
+`ClusterSelected -> Validated -> PrecacheSpecValid -> PrecachingSucceeded -> <TALM workflow continues>`.
+
+Detailed user and worklow steps are outlined below.
+
+1. User creates a CGU CR with the `preCaching` field set to `true` and prepares a ConfigMap (named
+`cluster-group-upgrade-overrides`) with the list of additional images to be pre-cached as shown in
+the examples below. Note that the ConfigMap should be created concurrently with the CGU or should
+exist prior to the creation of it. Furthermore, the ConfigMap should be created in the same namespace
+as the CGU on the hub cluster.
+   - Example CGU CR:
+    ```yaml
+    apiVersion: ran.openshift.io/v1alpha1
+    kind: ClusterGroupUpgrade
+    metadata:
+      name: cgu
+      namespace: default
+    spec:
+      clusters:
+      - sno1
+      - sno3
+      - sno7
+      enable: false
+      preCaching: true
+      backup: false
+      managedPolicies:
+      - policy0-common-config-policy
+      remediationStrategy:
+        maxConcurrency: 4
+        timeout: 240
+    ```
+   - Example ConfigMap:
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-group-upgrade-overrides
+      namespace: default
```

Comment:
Currently those additional images will be downloaded to all the clusters mentioned in the spec.Clusters, can we mention which specific images should be downloaded to which cluster, like more fine grained approach? Is it out of scope?

Reply:
Currently the CGU already allows the user to sub-select a set of clusters that is less than what the policy applies to. This allows pre-caching (and the policy rollout) to apply to a set of clusters of the user's choosing. Is there a scenario where you think further refinement would help?

Reply:
In case if we are upgrading various subset of clusters to different versions, i assume we need multiple cgu. For example, to upgrade 3 different subset of clusters will require 3 CGU and 3 configmap, considering every subset requires different image to be precached?

Reply:
Which images are pre-cached to which clusters can be controlled by the CGU itself with `spec.clusters` object. This implies that multiple CGUs can be created to support multiple pre-caching requirements at a granular level by setting the `spec.clusters` accordingly.